Implement a USB serial console

rjw57 opened the pull request:
Adds a stub USB serial console for the board. The most important commit is a12036d which adds a configuration option to `board.h` which is required to actually make USB work on our board. The other commits add a simple USB serial console based on the ChibiOS example code.

There are no commands beyond help, info, systime and exit now but one can actually connect to the board and ask it things over USB.

rjw57 committed:
firmware: disable OTG VBUS sense support

There are a plethora of ways to wire the STM32F4 up for USB OTG support. In our
case we are wanting peripheral-only mode and, importantly, without use of VBUS
sense. (This frees PA9 for other uses.)

Without BOARD_OTG_NOVBUSSENS defined, ChibiOS configures the MCU to use VBUS
sense support and since the PA9 line is probably unconnected, it will never
detect a USB cable as being plugged in.

rjw57 committed:
firmware: enable USB and USB serial

Add ChibiOS configuration to enable USB peripheral mode and USB serial console
driver.

rjw57 committed:
firmware: add USB configuration

Add USB configuration from ChibiOS STM32F4 example.

rjw57 committed:
firmware: add stub serial console

Add a simple serial console implementation. This just uses the stock ChibiOS
shell provider with no extra commands.

rjw57 committed:
firmware: enable serial console

Enable the serial console by launching the appropriate thread in main.c.

rjw57 commented:
Screenshot for, as is said on the Internet, pics or it never happened:

![screenshot from 2016-03-14 22-41-21](https://cloud.githubusercontent.com/assets/146860/13762299/e7534c8a-ea35-11e5-825e-4f7b3f730fc1.png)

@Tustvold: we probably don't want an interactive console in the final firmware but for the moment it might make debugging a hell of a lot easier!

adamgreig commented:
Check out the `threads` function from m2fc:
https://github.com/cuspaceflight/m2-electronics/blob/master/m2fc/firmware/m2fc_shell.c#L34

rjw57 committed:
firmware:serialconsole: add "threads" command

Add threads command based on [1].

[1] https://github.com/cuspaceflight/m2-electronics/blob/master/m2fc/firmware/m2fc_shell.c#L34

rjw57 commented:
Cool. I've tacked a version onto this PR since it looks highly useful.
